srt-protocol/src/packet/mod.rs

```diff
@@ -138,4 +135,53 @@ impl Debug for Packet {
     }
 }
 
-pub type ReceivePacketResult = Result<(Packet, SocketAddr), PacketParseError>;
+#[derive(Debug)]
+pub enum ReceivePacketError {
+    Parse(PacketParseError),
+    Io(io::Error),
+}
+
+impl From<io::Error> for ReceivePacketError {
+    fn from(error: io::Error) -> Self {
+        ReceivePacketError::Io(error)
+    }
+}
+
+impl From<PacketParseError> for ReceivePacketError {
+    fn from(error: PacketParseError) -> Self {
+        ReceivePacketError::Parse(error)
+    }
+}
+
+impl fmt::Display for ReceivePacketError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        <Self as fmt::Debug>::fmt(self, f)
+    }
+}
+
+impl std::error::Error for ReceivePacketError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        if let ReceivePacketError::Io(e) = self {
```

Comment:
I beleive we can always propagate, even for parse errors?

Reply:
Ok, well right now it is unclear which error value indicates that the underlying channel/socket has been closed. Since this struct is intended to abstract what could be a channel or socket, it needs to mash the semantics of the two together in a useful but clear way. Tough I'm not going to die on this hill, I don't wish to make the result type Result<Option<Packet>, RecievePacketError>, so my idea was to have RecievePacketError include end of stream semantics. This is actually a rough edge I want to get right before merge. Because it caused subtle deadlocking bug in the MulticastListener.